### PR TITLE
adding required annotations check and necessary tests

### PR DESCRIPTION
**Pull request by acornett21**
Adding a new check that checks for newly required annotations. At first we will only warn partners about this, then after sometime we will move to enforce this check.

- Relates: JIRA: EET-3540

**Note: There are backend changes being made to support `Warning` being in the results**

### CODE REVIEW

**Comment by tkrishtop** [comment]
Hi @acornett21 the DCI job failed because of a new test you've added
```
TASK [redhatci.ocp.verify_tests : Fail if not all test results are as expected for preflight_operator_simple-demo-operator_results-junit.xml]
0s
task path: /usr/share/ansible/collections/ansible_collections/redhatci/ocp/roles/verify_tests/tasks/parse_junit_file.yml:35
fatal: [jumphost]: FAILED! => {"changed": false, "msg": "The following expectations failed: [{'failed_expectation': {'testcase': '[a-zA-Z]+', 'passed': True}, 'actual_result': {'testcase': 'RequiredAnnotations', 'passed': False}}]"}
```
I'll change the expectations for simple-demo-operator tomorrow - for the moment we simply expect all tests to be green.

File: internal/formatters/util.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ func getResponse(r certification.Results) UserResponse {
 	passedChecks := make([]checkExecutionInfo, 0, len(r.Passed))
 	failedChecks := make([]checkExecutionInfo, 0, len(r.Failed))
 	erroredChecks := make([]checkExecutionInfo, 0, len(r.Errors))
+	warningChecks := make([]checkExecutionInfo, 0, len(r.Warning))
```
Comment (bcrochet) [inline internal/formatters/util.go:14]: To keep the tense, this should be `warnedChecks`

File: internal/formatters/util.go [hunk 2]
```diff
@@ -47,15 +48,30 @@ func getResponse(r certification.Results) UserResponse {
 		}
 	}
 
+	if len(r.Warning) > 0 {
```
Comment (bcrochet) [inline internal/formatters/util.go:51]: Same here... `Warned`

File: internal/policy/operator/required_annotations.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+func (h RequiredAnnotations) Name() string {
+	return "RequiredAnnotations"
+}
+
+func (h RequiredAnnotations) Metadata() check.Metadata {
+	return check.Metadata{
+		Description: "Checks that the CSV has all of the required feature annotations.",
+		// TODO: This will start as warning, but will need to move to `best`
+		Level:            "warning",
```
Comment (bcrochet) [inline internal/policy/operator/required_annotations.go:80]: Might go with `warn` instead of `warning`. I could be convinced otherwise on this one.

File: internal/formatters/util.go [hunk 4]
```diff
@@ -47,15 +48,30 @@ func getResponse(r certification.Results) UserResponse {
...
 		LibraryInfo:       version.Version,
 		CertificationHash: r.CertificationHash,
 		Results: resultsText{
-			Passed: passedChecks,
-			Failed: failedChecks,
-			Errors: erroredChecks,
+			Passed:  passedChecks,
+			Failed:  failedChecks,
+			Errors:  erroredChecks,
+			Warning: warningChecks,
```
Comment (bcrochet) [inline internal/formatters/util.go:74]:
```suggestion
			Warnings: warningChecks,
```

File: internal/formatters/util.go [hunk 5]
```diff
@@ -73,9 +89,10 @@ type UserResponse struct {
 
 // resultsText represents the results of check execution against the asset.
 type resultsText struct {
-	Passed []checkExecutionInfo `json:"passed" xml:"passed"`
-	Failed []checkExecutionInfo `json:"failed" xml:"failed"`
-	Errors []checkExecutionInfo `json:"errors" xml:"errors"`
+	Passed  []checkExecutionInfo `json:"passed" xml:"passed"`
+	Failed  []checkExecutionInfo `json:"failed" xml:"failed"`
+	Errors  []checkExecutionInfo `json:"errors" xml:"errors"`
+	Warning []checkExecutionInfo `json:"warning,omitempty" xml:"warning,omitempty"`
```
Comment (bcrochet) [inline internal/formatters/util.go:95]:
```suggestion
	Warnings []checkExecutionInfo `json:"warning,omitempty" xml:"warning,omitempty"`
```

File: internal/formatters/junitxml_test.go [hunk 6]
```diff
@@ -78,6 +78,38 @@ var _ = Describe("JUnitXML Formatter", func() {
 						ElapsedTime: 0,
 					},
 				},
+				Warning: []certification.Result{
```
Comment (bcrochet) [inline internal/formatters/junitxml_test.go:81]: If junitxml.go is not being modified, this seems odd to add. If you want to just make sure that the junit code isn't broken, then one of them should suffice.

**Review by bcrochet — CHANGES_REQUESTED** [review]
A few comments.

**Comment by acornett21** [comment]
> I'll change the expectations for simple-demo-operator tomorrow - for the moment we simply expect all tests to be green.

@tkrishtop Are you saying the new check showed up as failed? If so that's incorrect. This is still WIP, and I'll be working to update `simple-demo-operator` and `simple-disconnected-operator` accordingly, once this feature is fully hashed out.

File: internal/csv/csv.go [hunk 7]
```diff
@@ -10,7 +10,16 @@ import (
 	corev1 "k8s.io/api/core/v1"
 )
 
-const InfrastructureFeaturesAnnotation = "operators.openshift.io/infrastructure-features"
+const (
+	InfrastructureFeaturesAnnotation = "operators.openshift.io/infrastructure-features"
+	DisconnectedAnnotation           = "features.operators.openshift.io/disconnected"
+	FipsCompliantAnnotation          = "features.operators.openshift.io/fips-compliant"
```
Comment (komish) [inline internal/csv/csv.go:16]: NIT, don't yell at me! :P

```suggestion
	FIPSCompliantAnnotation          = "features.operators.openshift.io/fips-compliant"
```

File: internal/formatters/util.go (hunk 5 above)
Comment (komish) [reply internal/formatters/util.go:95]: Another note: I'm glad we're omitting this if it's empty, but we should make sure the backend doesn't expect it if we don't send it. (E.g. make sure it can be omitted there too). No action required here, just something to note to you since you're working with implementation on the backend

File: internal/policy/operator/required_annotations.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		KnowledgeBaseURL: "https://todo.com",
+		CheckURL:         "https://todo.com",
```
Comment (komish) [inline internal/policy/operator/required_annotations.go:82]: Just a reminder that this is a TODO!

File: internal/csv/csv.go (hunk 7 above)
Comment (bcrochet) [reply internal/csv/csv.go:16]: Not a nit at all to me. +1

File: internal/csv/csv.go [hunk 9]
```diff
@@ -10,7 +10,16 @@ import (
...
+	ProxyAwareAnnotation             = "features.operators.openshift.io/proxy-aware"
+	TLSProfilesAnnotation            = "features.operators.openshift.io/tls-profiles"
+	TokenAuthAWSAnnotation           = "features.operators.openshift.io/token-auth-aws"
+	TokenAuthAzureAnnotation         = "features.operators.openshift.io/token-auth-azure"
+	TokenAuthGCPAnnotation           = "features.operators.openshift.io/token-auth-gcp"
```
Comment (komish) [inline internal/csv/csv.go:21]: This is a potentially unrelated question.. but why are these becoming required? Are we installing operators on all providers all at once?

Comment (acornett21) [reply internal/csv/csv.go:21]: I'm not sure if a given operator would/could support token auth for all cloud providers or not. My guess would be that is unlikely, just based on our dealings with STS in the ACK project. IMO It's more so for the below:

> The current format of the infrastructure annotations is unsuited for robust checks and effective enforcement. We need a new format that allows to cleanly differentiate between mere absence of a data or actual lack of support for a certain infrastructure feature.

File: internal/policy/operator/required_annotations.go (hunk 3 above)
Comment (acornett21) [reply internal/policy/operator/required_annotations.go:80]: `warn` probably makes most sense, normal checks are level `best`

File: internal/formatters/junitxml_test.go (hunk 6 above)
Comment (acornett21) [reply internal/formatters/junitxml_test.go]: I wasn't really sure how this formatter worked based on the unit test, after testing against an actual cluster, I think there will be changes to the formatter, but I'm not sure what all is needed.

**Comment by tkrishtop** [comment]
I adjusted the pinned tests on our side, to ignore the new test failing for simple-demo. For the future, it would be nice to fix the simple-demo as well.

**Comment by tkrishtop** [comment]
check workload preflight-green

**Comment by acornett21** [comment]
@tkrishtop I've updated the the `junit` code to add another level/tag (not sure what to call this for junit). But the output will look like this for checks that are at a `warn` level.

```
<testsuites>
	<testsuite tests="5" failures="0" warnings="1" time="35.215827" name="Red Hat Certification">
		<properties></properties>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="ScorecardBasicSpecCheck" time="3.516921">Check to make sure that all CRs have a spec block.</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="ScorecardOlmSuiteCheck" time="3.530103">Operator-sdk scorecard OLM Test Suite Check</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="DeployableByOLM" time="28.148170">Checking if the operator could be deployed by OLM</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="ValidateOperatorBundle" time="0.010793">Validating Bundle image that checks if it can validate the content and format of the operator bundle</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="RequiredAnnotations" time="9.839866ms">
			<warning message="Warn" type="">Check that the CSV has all of the required feature annotations.: Suggested Fix: Add all of the required annotations, and make sure the value is set to either &#39;true&#39; or &#39;false&#39;</warning>
		</testcase>
	</testsuite>
</testsuites>
```

The idea of introducing a new check at level `warn` is so that partners can see the check in their results, but not be penalized if it fails. Then at a later date the check will be moved to level `best` where the check is enforced. I hope this clears up what we are trying to accomplish. Let me know if we need to adjust the XML, or if you see any issue in how the XML works, since I think DCI is the only consumer of this.

**Comment by acornett21** [comment]
Tested locally as well now that the operators are updated...
```
{
    "image": "quay.io/opdev/simple-demo-operator-bundle:v0.0.7",
    "passed": true,
    "certification_hash": "d41d8cd98f00b204e9800998ecf8427e",
    "test_library": {
        "name": "github.com/redhat-openshift-ecosystem/openshift-preflight",
        "version": "unknown",
        "commit": "unknown"
    },
    "results": {
        "passed": [
            {
                "name": "ScorecardBasicSpecCheck",
                "elapsed_time": 12673,
                "description": "Check to make sure that all CRs have a spec block."
            },
            {
                "name": "ScorecardOlmSuiteCheck",
                "elapsed_time": 3507,
                "description": "Operator-sdk scorecard OLM Test Suite Check"
            },
            {
                "name": "DeployableByOLM",
                "elapsed_time": 66839,
                "description": "Checking if the operator could be deployed by OLM"
            },
            {
                "name": "ValidateOperatorBundle",
                "elapsed_time": 12,
                "description": "Validating Bundle image that checks if it can validate the content and format of the operator bundle"
            },
            {
                "name": "RequiredAnnotations",
                "elapsed_time": 11,
                "description": "Checks that the CSV has all of the required feature annotations."
            }
        ],
        "failed": [],
        "errors": []
    }
}
```

File: internal/engine/engine.go [hunk 10]
```diff
@@ -236,6 +236,12 @@ func (c *craneEngine) ExecuteChecks(ctx context.Context) error {
 		}
 
 		if !checkPassed {
+			// if a test doesn't pass but is of level warn include it in warning results, instead of failed results
+			if check.Metadata().Level == "warn" {
```
Comment (bcrochet) [inline internal/engine/engine.go:240]: This is different... You use "warning" on 223, but "warn" here. Perhaps time for a constant for these levels?

File: internal/engine/engine_test.go [hunk 11]
```diff
@@ -108,6 +108,24 @@ var _ = Describe("Execute Checks tests", func() {
 			check.HelpText{},
 		)
 
+		warningCheckPassing := check.NewGenericCheck(
+			"warnCheckPassing",
+			func(context.Context, image.ImageReference) (bool, error) {
+				return true, nil
+			},
+			check.Metadata{Level: "warn"},
```
Comment (bcrochet) [inline internal/engine/engine_test.go:116]: These strings are being used enough that a constant is warranted.

File: internal/engine/engine_test.go [hunk 12]
```diff
@@ -108,6 +108,24 @@ var _ = Describe("Execute Checks tests", func() {
...
+			check.HelpText{},
+		)
+
+		warningCheckFailing := check.NewGenericCheck(
+			"warnCheckFailing",
+			func(context.Context, image.ImageReference) (bool, error) {
+				return false, nil
+			},
+			check.Metadata{Level: "warn"},
```
Comment (bcrochet) [inline internal/engine/engine_test.go:125]: Ditto.

File: internal/policy/operator/required_annotations_test.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+		})
+
+		It("Should successfully find a CSV in a valid bundle", func() {
+			csv, err := h.getBundleCSV(context.TODO(), "./testdata/disconnected_bundle")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(csv.ObjectMeta.Name).ToNot(BeEmpty())
+		})
+	})
+
+	When("Validating that a CSV has all of the required annotations.", func() {
```
Comment (bcrochet) [inline internal/policy/operator/required_annotations_test.go:31]: Nit: Remove the '.'

File: internal/policy/operator/restricted_network_aware.go [hunk 14]
```diff
@@ -37,6 +37,8 @@ func (p FollowsRestrictedNetworkEnablementGuidelines) validate(ctx context.Conte
 		return false, err
 	}
 
+	// TODO: This annotation is going to be DEPRECATED, so we should also check for the new `disconnected` specified annotation
```
Comment (bcrochet) [inline internal/policy/operator/restricted_network_aware.go:40]: Unrelated change.

**Review by bcrochet — CHANGES_REQUESTED** [review]
A few requests.

File: internal/policy/operator/restricted_network_aware.go (hunk 14 above)
Comment (acornett21) [reply internal/policy/operator/restricted_network_aware.go:40]: This is related...when the required annotations check gets enforced, this restricted check will have to be updated to the new annotation format. Is it just a wording problem? I can re-word this to be clearer.

Comment (acornett21) [reply internal/policy/operator/restricted_network_aware.go:40]: After thinking about it, it probably makes more sense for this to live in a GH/JIRA issues for all the things that need to change once the required annotations check is enforced.

File: internal/engine/engine.go (hunk 10 above)
Comment (acornett21) [reply internal/engine/engine.go]: I moved this to a constant, though I'm not sure the package is the ideal place. I put it in `check.certification` which defines the `Check` interface.

File: internal/formatters/util.go (hunk 5 above)
Comment (acornett21) [reply internal/formatters/util.go]: This is optional in the [API spec](https://catalog.redhat.com/api/containers/v1/ui/#/Certification%20projects/graphql.cert_project_test_result.create_certification_test_result).

File: internal/formatters/junitxml_test.go (hunk 6 above)
Comment (acornett21) [reply internal/formatters/junitxml_test.go]: This is where we landed on the formatter format for warnings:

```
<testsuites>
	<testsuite tests="5" failures="0" warnings="1" time="32.611190" name="Red Hat Certification">
		<properties></properties>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="ScorecardBasicSpecCheck" time="2.641964">Check to make sure that all CRs have a spec block.</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="ScorecardOlmSuiteCheck" time="3.491450">Operator-sdk scorecard OLM Test Suite Check</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="DeployableByOLM" time="26.456957">Checking if the operator could be deployed by OLM</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="ValidateOperatorBundle" time="0.011117">Validating Bundle image that checks if it can validate the content and format of the operator bundle</testcase>
		<testcase classname="quay.io/opdev/simple-demo-operator-bundle:v0.0.6" name="RequiredAnnotations" time="9.701301ms">
			<warning message="Warn" type="">Check that the CSV has all of the required feature annotations.: Suggested Fix: Add all of the required annotations, and make sure the value is set to either &#39;true&#39; or &#39;false&#39;</warning>
		</testcase>
	</testsuite>
</testsuites>
```

File: internal/policy/operator/restricted_network_aware.go (hunk 14 above)
Comment (acornett21) [reply internal/policy/operator/restricted_network_aware.go]: Created the below to cover this and removed the `TODO`
- #1104

File: internal/engine/engine.go [hunk 15]
```diff
@@ -23,7 +23,7 @@ import (
 
 	"github.com/redhat-openshift-ecosystem/openshift-preflight/artifacts"
 	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
-	"github.com/redhat-openshift-ecosystem/openshift-preflight/internal/check"
+	certcheck "github.com/redhat-openshift-ecosystem/openshift-preflight/internal/check"
```
Comment (komish) [inline internal/engine/engine.go:26]: NIT: Why is this changing?

File: internal/engine/engine.go [hunk 16]
```diff
@@ -220,7 +220,7 @@ func (c *craneEngine) ExecuteChecks(ctx context.Context) error {
 		c.results.TestedImage = c.image
 
 		logger.V(log.DBG).Info("running check", "check", check.Name())
-		if check.Metadata().Level == "optional" {
+		if check.Metadata().Level == "optional" || check.Metadata().Level == certcheck.LevelWarn {
```
Comment (bcrochet) [inline internal/engine/engine.go:223]: Why not use the Optional const?

File: internal/engine/engine.go (hunk 15 above)
Comment (bcrochet) [reply internal/engine/engine.go:26]: Agreed. This is an awkward change.

File: internal/engine/engine.go [hunk 17]
```diff
@@ -220,7 +220,7 @@ func (c *craneEngine) ExecuteChecks(ctx context.Context) error {
 		c.results.TestedImage = c.image
```
Comment (bcrochet) [inline internal/engine/engine.go:220]: Assuming you make the change to the import because of the name of this very temporary variable. I'd rather see this change rather than the import.

File: internal/engine/engine.go [hunk 18]
```diff
@@ -743,7 +750,7 @@ func makeCheckList(checks []check.Check) []string {
 
 // checkNamesFor produces a slice of names for checks in the requested policy.
 func checkNamesFor(ctx context.Context, p policy.Policy) []string {
-	var c []check.Check
+	var c []certcheck.Check
```
Comment (bcrochet) [inline internal/engine/engine.go:753]: Yeah. Definitely do not want this. Way too many unnecessary, unrelated changes.

**Review by bcrochet — CHANGES_REQUESTED** [review]
Notes about the import name

File: internal/engine/engine.go (hunk 15 above)
Comment (acornett21) [reply internal/engine/engine.go:26]: This is due to scoping, I can't reference the `check` package in a scoped for loop on line 223, since `check` is a struct there. If there is a better way to go about this LMK.

File: internal/engine/engine.go (hunk 17 above)
Comment (komish) [reply internal/engine/engine.go:220]: Agreed - change this and leave the import. It's a smaller change.

File: internal/engine/engine.go (hunk 16 above)
Comment (acornett21) [reply internal/engine/engine.go:223]: To keep this PR focused on the required annotations, I was planning a follow up PR to switch all other checks to the constants. I assumed that would be preference then introducing here. But maybe it's fine to just introduce for this one line since it is already changing?

File: internal/engine/engine.go (hunk 17 above)
Comment (acornett21) [reply internal/engine/engine.go:220]: It's actually not for this, it's for the code on line 223, since there `check` is a scoped struct as part of a for loop.

**Review by komish — APPROVED** [review]
Barring E2E, LGTM.

/lgtm

File: internal/engine/engine.go (hunk 17 above)
Comment (bcrochet) [reply internal/engine/engine.go:220]: I was referring to the name of the var in the for loop.

**Review by bcrochet — APPROVED** [review]
/lgtm
